storage/indexer/src/db_tailer.rs

```diff
@@ -0,0 +1,296 @@
+    pub db: Arc<DB>,
+    pub main_db_reader: Arc<dyn DbReader>,
+    batch_size: usize,
+}
+
+impl DBTailer {
+    pub fn new(db: Arc<DB>, db_reader: Arc<dyn DbReader>, config: &IndexDBTailerConfig) -> Self {
+        Self {
+            db,
+            main_db_reader: db_reader,
+            batch_size: config.batch_size,
+        }
+    }
+
+    pub fn get_persisted_version(&self) -> Version {
+        // read the latest key from the db
+        let mut rev_iter_res = self
+            .db
+            .rev_iter::<TailerMetadataSchema>(Default::default())
+            .expect("Cannot create db tailer metadata iterator");
+        rev_iter_res
+            .next()
+            .map(|res| res.map_or(0, |(version, _)| version))
+            .unwrap_or_default()
+    }
+
+    pub fn process_a_batch(&self, start_version: Option<Version>) -> Result<Version> {
+        let mut version = start_version.unwrap_or(0);
+        let db_iter: Box<
+            dyn Iterator<
+                Item = std::prelude::v1::Result<
+                    (aptos_types::transaction::Transaction, Vec<ContractEvent>),
+                    AptosDbError,
+                >,
+            >,
+        > = self
+            .main_db_reader
+            .get_db_tailor_iter(version, self.batch_size)
+            .expect("Cannot create db tailer iterator");
+        let batch = SchemaBatch::new();
+        db_iter.for_each(|res| {
+            res.map(|(txn, events)| {
+                if let Some(txn) = txn.try_as_signed_user_txn() {
+                    batch
+                        .put::<TransactionByAccountSchema>(
+                            &(txn.sender(), txn.sequence_number()),
+                            &version,
+                        )
+                        .expect("Failed to put txn to db tailer batch");
+
+                    events.iter().enumerate().for_each(|(idx, event)| {
+                        if let ContractEvent::V1(v1) = event {
+                            batch
+                                .put::<EventByKeySchema>(
+                                    &(*v1.key(), v1.sequence_number()),
+                                    &(version, idx as u64),
+                                )
+                                .expect("Failed to event by key to db tailer batch");
+                            batch
+                                .put::<EventByVersionSchema>(
+                                    &(*v1.key(), version, v1.sequence_number()),
+                                    &(idx as u64),
+                                )
+                                .expect("Failed to event by version to db tailer batch");
+                        }
+                    });
+                }
+                version += 1;
+            })
+            .expect("Failed to iterate db tailer iterator");
+        });
+        batch.put::<TailerMetadataSchema>(&version, &())?;
+        self.db.write_schemas(batch)?;
+        Ok(version)
```

Comment:
Shall we sleep 10ms, for example, if it didn't see any new txns? (Or did you do it somewhere else?)

Reply:
and shall we make it a pipeline (building the batch and committing it) from the beginning?

Reply:
It sleeps at the tailer service `run` function if no new data from db iterator. made it pipeline